src/platform/recording.py

```diff
@@ -0,0 +1,183 @@
+
+
+class RequestRecorder:
+    """SQLite-backed recorder for request/response interactions.
+
+    Thread-safe; maintains a single connection. Designed for tests and lightweight usage.
+    """
+
+    def __init__(self, db_path: Path | None = None):
+        self.db_path = Path(db_path) if db_path else DEFAULT_DB_PATH
+        _ensure_parent_dir(self.db_path)
+        # Use a single connection; enable WAL for better concurrency
+        self._conn = sqlite3.connect(str(self.db_path), check_same_thread=False)
+        self._conn.execute("PRAGMA journal_mode=WAL;")
+        self._conn.execute("PRAGMA synchronous=NORMAL;")
+        self._lock = threading.Lock()
+        self._init_schema()
+
+    def _init_schema(self) -> None:
+        with self._conn:  # implicit transaction
+            self._conn.execute(
+                """
+                CREATE TABLE IF NOT EXISTS interactions (
+                    id INTEGER PRIMARY KEY,
+                    timestamp TEXT NOT NULL,
+                    method TEXT NOT NULL,
+                    request_json TEXT NOT NULL,
+                    request_hash TEXT NOT NULL,
+                    sequence_num INTEGER NOT NULL,
+                    response_json TEXT NOT NULL,
+                    duration_ms INTEGER NOT NULL
+                );
+                """
+            )
+            self._conn.execute(
+                """
+                CREATE UNIQUE INDEX IF NOT EXISTS idx_interactions_unique
+                ON interactions(request_hash, sequence_num);
+                """
+            )
+            self._conn.execute(
+                """
+                CREATE INDEX IF NOT EXISTS idx_interactions_method
+                ON interactions(method);
+                """
+            )
+
+    def record(
+        self,
+        *,
+        method: str,
+        kwargs: dict,
+        response: Any,
+        duration_ms: int,
+        db_path: Path | None = None,
+    ) -> None:
+        """Record a successful interaction.
+
+        Args:
+            method: Fully qualified method name.
+            kwargs: Request arguments.
+            response: Successful response object (JSON-serializable or encodable).
+            duration_ms: Duration of the call in milliseconds.
+            db_path: Optional override DB path for this call.
+        """
+
+        if db_path and Path(db_path) != self.db_path:
+            # If an override is provided, write to that DB lazily via a separate recorder
+            override = RequestRecorder(Path(db_path))
+            override.record(
+                method=method, kwargs=kwargs, response=response, duration_ms=duration_ms
+            )
+            return
+
+        request_json = json.dumps(kwargs, sort_keys=True, separators=(",", ":"))
+        request_hash = compute_request_hash(method, kwargs)
+        response_json = json.dumps(response, sort_keys=True, separators=(",", ":"))
+        now_iso = time.strftime("%Y-%m-%dT%H:%M:%S", time.gmtime())
```

Comment:
Consider `time.time_ns()` or `datetime.datetime.utcnow().isoformat(timespec="microseconds")` for ordering during rapid tests.